src/eval/mod.rs

```diff
@@ -567,12 +577,24 @@ where
                                     let mut thunk = env.get(var_id).ok_or_else(|| {
                                         EvalError::UnboundIdentifier(var_id.clone(), pos)
                                     })?;
-
-                                    thunk.borrow_mut().env.extend(
-                                        rec_env
-                                            .iter_elems()
-                                            .map(|(id, thunk)| (id.clone(), thunk.clone())),
-                                    );
+                                    let mut clos = thunk.borrow_mut();
+                                    if let Some(Some(free_vars)) = free_vars
+                                        .as_ref()
+                                        .map(|fr| fr.get(&Ident::from(String::new())))
```

Comment:
Nitpick: you can avoid the double `Some` by using `and_then` (aka flatmap) instead of `map`.